src/Sarif.Multitool.Library/MergeCommand.cs

```diff
@@ -17,142 +21,251 @@ public class MergeCommand : CommandBase
-                string outputFilePath = Path.Combine(outputDirectory, GetOutputFileName(mergeOptions));
+                string outputFilePath = Path.Combine(outputDirectory, GetOutputFileName(_options));
 
-                if (mergeOptions.SplittingStrategy == 0)
+                if (_options.SplittingStrategy == 0)
                 {
-                    if (!DriverUtilities.ReportWhetherOutputFileCanBeCreated(outputFilePath, mergeOptions.Force, _fileSystem))
+                    if (!DriverUtilities.ReportWhetherOutputFileCanBeCreated(outputFilePath, _options.Force, _fileSystem))
                     {
                         return FAILURE;
                     }
                 }
 
-                HashSet<string> sarifFiles = CreateTargetsSet(mergeOptions.TargetFileSpecifiers, mergeOptions.Recurse, _fileSystem);
-
-                IEnumerable<SarifLog> allRuns = ParseFiles(sarifFiles);
+                var logLoadOptions = new BoundedChannelOptions(1000)
+                {
+                    SingleWriter = true,
+                    SingleReader = false,
+                };
+                _logLoadChannel = Channel.CreateBounded<string>(logLoadOptions);
 
-                // Build one SarifLog with all the Runs.
-                SarifLog mergedLog = allRuns
-                    .Merge(mergeOptions.MergeEmptyLogs)
-                    .InsertOptionalData(mergeOptions.DataToInsert.ToFlags())
-                    .RemoveOptionalData(mergeOptions.DataToInsert.ToFlags());
+                var mergeLogsOptions = new UnboundedChannelOptions()
+                {
+                    SingleWriter = false,
+                    SingleReader = true
+                };
+                _mergeLogsChannel = Channel.CreateUnbounded<SarifLog>(mergeLogsOptions);
 
-                // If there were no input files, the Merge operation set combinedLog.Runs to null. Although
-                // null is valid in certain error cases, it is not valid here. Here, the correct value is
-                // an empty list. See the SARIF spec, §3.13.4, "runs property".
-                mergedLog.Runs ??= new List<Run>();
-                mergedLog.Version = SarifVersion.Current;
-                mergedLog.SchemaUri = mergedLog.Version.ConvertToSchemaUri();
+                _options.Threads = _options.Threads > 0 ? _options.Threads : Environment.ProcessorCount;
 
-                if (mergeOptions.SplittingStrategy != SplittingStrategy.PerRule)
+                var workers = new Task<bool>[_options.Threads];
+                for (int i = 0; i < _options.Threads; i++)
                 {
-                    _fileSystem.DirectoryCreate(outputDirectory);
-
-                    // Write output to file.
-                    WriteSarifFile(_fileSystem, mergedLog, outputFilePath, mergeOptions.Formatting);
-                    return 0;
+                    workers[i] = Task.Run(LoadSarifLogs);
```

Comment:
I think I found the problem. Right now, we are creating multiple readers and I think it wasn't designed for that. Looking at the documentation, I saw the following example:
1 reader and 1 writer
1 reader and multiple writers

Reply:
fixed. ignore my comment.